modyn/storage/include/internal/grpc/storage_service_impl.hpp

```diff
@@ -186,6 +186,138 @@ class StorageServiceImpl final : public modyn::storage::Storage::Service {
   }
 
+  template <typename WriterT>
+  Status GetDataPerWorker_Impl(  // NOLINT (readability-identifier-naming)
+      ServerContext* /*context*/, const modyn::storage::GetDataPerWorkerRequest* request, WriterT* writer) {
+    try {
+      soci::session session = storage_database_connection_.get_session();
+
+      // Check if the dataset exists
+      int64_t dataset_id = get_dataset_id(session, request->dataset_id());
+
+      if (dataset_id == -1) {
+        SPDLOG_ERROR("Dataset {} does not exist.", request->dataset_id());
+        return {StatusCode::OK, "Dataset does not exist."};
+      }
+
+      const int64_t start_timestamp = request->start_timestamp();
+      const int64_t end_timestamp = request->end_timestamp();
+
+      SPDLOG_INFO(
+          fmt::format("Received GetDataPerWorker Request for dataset {} (id = {}) and worker {} out of {} workers with"
+                      " start = {} and end = {}.",
+                      request->dataset_id(), dataset_id, request->worker_id(), request->total_workers(),
+                      start_timestamp, end_timestamp));
+
+      int64_t total_keys = 0;
+      std::string timestamp_filter;
+      if (start_timestamp > 0 && end_timestamp == 0) {
+        timestamp_filter = fmt::format("updated_at >= {}", start_timestamp);
+      } else if (start_timestamp == 0 && end_timestamp > 0) {
+        timestamp_filter = fmt::format("updated_at <= {}", end_timestamp);
+      } else if (start_timestamp > 0 && end_timestamp > 0) {
+        timestamp_filter = fmt::format("updated_at >= {} AND updated_at <= {}", start_timestamp, end_timestamp);
+      } else if (start_timestamp == 0 && end_timestamp == 0) {
+        timestamp_filter = "1 = 1";
+      } else {
+        FAIL(fmt::format("Invalid timestamps: start = {}, end = {}", start_timestamp, end_timestamp));
+      }
+
+      session << "SELECT COALESCE(SUM(number_of_samples), 0) FROM files WHERE dataset_id = :dataset_id AND " +
+                     timestamp_filter,
+          soci::into(total_keys), soci::use(dataset_id);
+
+      if (total_keys > 0) {
+        int64_t start_index = 0;
+        int64_t limit = 0;
+        std::tie(start_index, limit) =
+            get_partition_for_worker(request->worker_id(), request->total_workers(), total_keys);
+
+        std::string query;
+        if (start_timestamp == 0 && end_timestamp == 0) {
+          query =
+              fmt::format("SELECT sample_id FROM samples WHERE dataset_id = {} ORDER BY sample_id LIMIT {} OFFSET {}",
+                          dataset_id, limit, start_index);
+        } else {
+          query = fmt::format(
+              "SELECT samples.sample_id "
+              "FROM samples INNER JOIN files "
+              "ON samples.file_id = files.file_id AND samples.dataset_id = files.dataset_id "
+              "WHERE samples.dataset_id = {} AND {} "
+              "ORDER BY sample_id LIMIT {} OFFSET {}",
+              dataset_id, timestamp_filter, limit, start_index);
```

Comment:
Since `GetDataPerWorker` API is only used in evaluation, which is off the critical path of training.
I consider this as okay

Reply:
the `send_sample_id_and_label` implementation does the same so should be ok, however, I still wonder whether we can deduplicate the code somehow.

Reply:
Yes correct! I noticed it as well. I didn't merge the implementation because
1. in `send_sample_id_and_label` it is ordered by `updated_at` and here by `sample_id`.
2. the response of `sned_sample_id_and_label` is not sharded by `worker_id`.

It is possible to merge the implementation but considering the logic itself is less than 10 lines, and the merge has to take care of the above differences (not a trivial merge), so I thought it is not worth the effort.